README.rst
===========
Workalendar
===========

Overview
========

Workalendar is a Python module that offers classes able to handle calendars,
list legal / religious holidays and gives working-day-related computation
functions.


Status
======

This is barely beta. Please consider this module as a work in progres.


Usage sample
============

.. code-block:: python

    >>> from datetime import date
    >>> from workalendar.europe import France
    >>> cal = France()
    >>> cal.holidays(2012)
    [(datetime.date(2012, 1, 1), 'New year'),
     (datetime.date(2012, 4, 9), 'Easter Monday'),
     (datetime.date(2012, 5, 1), 'Labour Day'),
     (datetime.date(2012, 5, 8), 'Victory in Europe Day'),
     (datetime.date(2012, 5, 17), 'Ascension Day'),
     (datetime.date(2012, 5, 28), 'Whit Monday'),
     (datetime.date(2012, 7, 14), 'Bastille Day'),
     (datetime.date(2012, 8, 15), 'Assumption of Mary to Heaven'),
     (datetime.date(2012, 11, 1), "All Saints' Day"),
     (datetime.date(2012, 11, 11), 'Armistice Day'),
     (datetime.date(2012, 12, 25), 'Christmas')]
    >>> cal.is_working_day(date(2012, 12, 25))  # it's Christmas
    False
    >>> cal.is_working_day(date(2012, 12, 30))  # it's Sunday
    False
    >>> cal.is_working_day(date(2012, 12, 26))
    True

External dependencies
=====================

You may want to install ``python-dev`` and/or ``python3-dev`` on your machine to
either run the installation or run tests via tox.


Tests
=====

Travis status:

.. image:: https://api.travis-ci.org/novapost/workalendar.png


To run test, just install tox with ``pip install tox`` and run::

    tox

from the command line.


Available Calendars
===================

Europe
------

* Czech Republic
* European Central Bank
* Finland
* France
* France (Alsace / Moselle)
* Greece
* Hungary
* Iceland
* Italy
* Norway
* Poland
* Sweden
* United Kingdom (incl. Northern Ireland, Scotland)

America
-------

* Brazil (incl. São Paulo state and city)
* Chile
* Mexico
* Panama
* United States of America (only the federal state holidays at the moment)

Asia
----

* Japan
* Qatar
* South Korea
* Taiwan

Oceania
-------

* Australia (incl. its different states)
* Marshall Islands

Africa
------

* Algeria
* Benin
* Ivory Coast
* Madagascar
* São Tomé
* South Africa

And more to come (I hope!)

Caveats
=======

Please take note that some calendars are not 100% accurate. The most common
example is the Islamic calendar, where some computed holidays are not exactly on
the same official day decided by religious authorities, and this may vary
country by country. Whenever it's possible, try to adjust your results with
the official data provided by the adequate authorities.

Contributing
============

Please read our `CONTRIBUTING.rst <https://github.com/novapost/workalendar/blob/master/CONTRIBUTING.rst>`_
document to discover how you can contribute to ``workalendar``. Pull-requests
are very welcome.

License
=======

This library is published under the terms of the MIT License. Please check the
LICENSE file for more details.


CHANGELOG
CHANGELOG
=========

master (unreleased)
-------------------

- Add the `get_working_days_delta` method to the core calendar class (#260).

2.4.0 (2018-03-28)
------------------

New Calendars
~~~~~~~~~~~~~

- Added Lithuania, by @landler (#254).
- Added Russia, by @vanadium23 (#259).

Bugfixes
~~~~~~~~

- Fixed shifting ANZAC day for Australia states (#249).
- Renamed Australian state classes to actual state names(eg. AustraliaNewSouthWales to NewSouthWales).
- Update ACT holidays (#251).
- Fixing Federal Christmas Shift ; added a `include_veterans_day` flag to enable/disable Veteran's day on specific calendar - e.g. Mozilla's dedicated calendar (#242).

Other
~~~~~

- **Deprecation:** Dropped support for Python 3.3 (#245).
- Fixed Travis-ci configuration for Python 3.5 and al (#252).
- Moved from `novafloss` to the `peopledoc` organization, the core People Doc Inc. organization (#255).
- First step iteration on the "global registry" feature. European countries are now part of a registry loaded in the ``workalenda.registry`` module. Please use with care at the moment (#248).
- Refactored Australia family and community day calculation (#244).

2.3.1 (2017-07-27)
------------------

I have done a terrible mistake with the 1.3.0 release, and uploaded a defunct 2.3.0 version... I tried to remove it from PyPI, but it failed at some point. In order to make sure that the latest version has the highest version number, I'm releasing this "dumb" version. It is equivalent to the 2.1.0 release, but at least, if you upgrade, you have all the 2.0+ niceties...

2.1.0 (2017-07-27)
------------------

- Added Hong Kong, by @nedlowe (#235).
- Splitted `africa.py` file into an `africa/` module (#236).
- Added Alabama Counties - Baldwin County, Mobile County, Perry County. Refactored UnitedStates classes to have a parameter to include the "Mardi Gras" day (#214).
- Added brazilian calendar to consider working days for bank transactions, by @fvlima (#238).

2.0.0 (2017-06-23)
------------------

- Major refactor in the USA module. Each State is now an independant module, all of the Mixins were removed, all the possible corrections have been made, following the main Wikipedia page, and cross-checking with official sources when it was possible (#171).
- Added District of Columbia in the USA module (#217).
- Run tests with Python3.6 in CI (#210)
- Small refactors / cleanups in the following calendars: Hungary, Iceland, Ireland, Latvia, Netherlands, Spain, Japan, Taiwan, Australia, Canada, USA (#209).
- Various refactors for the Asia module, essentially centered around a more convenient Chinese New Year computation toolset (#202).
- Refactoring the USA tests: using inheritance to test federal and state-based holidays using only one "Don't Repeat Yourself" codebase (#213).

1.3.0 (2017-06-09)
------------------

- Added Singapore calendar, initiated by @nedlowe (#194 + #195).
- Added Malaysia, by @gregyhj (#201).
- Added Good Friday in the list of Hungarian holidays, as of the year 2017 (#203), thx to @mariusz-korzekwa for the bug report.
- Assigned a minimal setuptools version, to avoid naughty ``DistributionNotFound`` exceptions with obsolete versions (#74).
- Fixed a bug in Slovakia calendar, de-duplicated Christmas Day, that appeared twice (#205).
- Fixed important bugs in the calendars of the following Brazilian cities: Vitória, Vila Velha, Cariacica, Guarapari and Serra - thx to Fernanda Gonçalves Rodrigues, who confirmed this issue raised by @Skippern (#199).

1.2.0 (2017-05-30)
------------------

- Moved all the calendar of countries on the american continent in their own modules (#188).
- Refactor base Calendar class get_weekend_days to use WEEKEND_DAYS more intelligently (#191 + #192).
- Many additions to the Brazil and various states / cities. Were added: Acre, Alagoas, Amapá, Amazonas, Bahia, Ceará, Distrito Federal, Espírito Santo State, Goiás, Maranhão, Mato Grosso, Mato Grosso do Sul, Pará, Paraíba, Pernambuco, Piauí, Rio de Janeiro, Rio Grande do Norte, Rio Grande do Sul, Rondônia, Roraima, Santa Catarina, São Paulo, Sergipe, Tocantins, City of Vitória, City of Vila Velha, City of Cariacica, City of Guarapari and City of Serra (#187).
- Added a ``good_friday_label`` class variable to ``ChristianMixin`` ; one can assign the right label to this holiday (#187).
- Added a ``ash_wednesday_label`` class variable to ``ChristianMixin`` ; one can assign the right label to this holiday (#187).


1.1.0 (2017-02-28)
------------------

- Added Cyprus. thx @gregn610 (#174).
- Added Latvia. thx @gregn610 (#178).
- Added Malta. thx @gregn610 (#179).
- Added Romania. thx @gregn610 (#180).
- Added Canton of Vaud (Switzerland) - @brutasse (#182).
- Fixed January 2nd state holiday (#181).
- Fixed Saxony repentance day for the year 2016. thx @Natim (#168).
- Fixed Historical and one-off holidays for South Africa. thx @gregn610 (#173).
- Minor PEP8 fixes (#186).


1.0.0 (2016-12-12)
------------------

After several years of development, we can now say that this library is production-ready, so we're releasing its 1.0.0 version. Millions of "thank you" to all the contributors involved.

- Add Ireland. thx @gregn610 (#152).
- Bugfix: New Year's Eve is not a holiday in Netherlands (#154).
- Add Austria.  thx @gregn610 (#153)
- Add Bulgaria. thx @gregn610 (#156)
- Add Croatia. thx @gregn610 (#157)


0.8.1 (2016-11-08)
------------------

- Reformation Day is a national holiday in Germany, but only in 2017 (#150).


0.8.0 (2016-08-25)
------------------

- Fix Czech Republic calendar - as of 2016, Good Friday has become a holiday (#148).


0.7.0 (2016-08-02)
------------------

- Easter Sunday is a Brandenburg federate state holiday (#143), thx @uvchik.
- Added Catalonia (#145), thx @ferranp.
- Use `find_packages()` to fetch package directories in `setup.py` (#141, #144).
- use py.test instead of nosetests for tests (#146).
- cleanup: remove unused ``swiss.py`` file (#147).

0.6.1 (2016-06-29)
------------------


workalendar/tests/test_scotland.py
from datetime import date

from workalendar.tests import GenericCalendarTest
from workalendar.europe import (
    Scotland,
    ScotlandAngus,
    ScotlandAyr,
    ScotlandCarnoustie,
    ScotlandClydebank,
    ScotlandDumfriesGalloway,
    ScotlandDundee,
    ScotlandEastDunbartonshire,
    ScotlandEdinburgh,
    ScotlandElgin,
    ScotlandFalkirk,
    ScotlandFife,
    ScotlandGalashiels,
    ScotlandGlasgow,
    ScotlandInverclyde,
    ScotlandInverness,
    ScotlandKilmarnock,
    ScotlandLochaber,
    ScotlandMonifieth,
    ScotlandPaisley,
    ScotlandPerth,
    ScotlandScottishBorders,
    ScotlandStirling,
    ScotlandSouth,
    ScotlandSouthLanarkshire,
    ScotlandNorthLanarkshire,
    ScotlandWestDunbartonshire,
)


# Test mixins
class GoodFridayMixin(object):

    def test_good_friday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 4, 18), holidays)  # Good friday


class EasterMondayMixin(object):

    def test_easter_monday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 4, 21), holidays)  # Easter monday


class SpringHolidayFirstMondayAprilMixin(object):

    def test_spring_holiday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 4, 7), holidays)  # Spring holiday


class SpringHolidaySecondMondayAprilMixin(object):

    def test_spring_holiday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 4, 14), holidays)  # Spring holiday


class SpringHolidayTuesdayMondayMayMixin(object):
    "Test the 'Tuesday after 1st Monday in May' rule"
    def test_spring_holiday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 5, 5), holidays)

    def test_spring_holiday_2012(self):
        # Special case: since May 1st is a Tuesday, it's not the 1st Tuesday
        holidays = self.cal.holidays_set(2012)
        self.assertIn(date(2012, 5, 8), holidays)


class VictoriaDayLastMondayBefore24MayMixin(object):
    "Test the 'Last Monday strictly before 24 May' rule"
    def test_victoria_day_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 5, 19), holidays)

    def test_victoria_day_2010(self):
        # Special case: May 24th is a Monday, but it's *strictly* this day
        holidays = self.cal.holidays_set(2010)
        self.assertIn(date(2010, 5, 17), holidays)


class SpringHolidayLastMondayMayMixin(object):
    def test_spring_holiday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 5, 26), holidays)

    def test_spring_holiday_2011(self):
        holidays = self.cal.holidays_set(2011)
        self.assertIn(date(2011, 5, 30), holidays)


class VictoriaDayLastMondayMayMixin(object):
    def test_victoria_day_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 5, 26), holidays)

    def test_victoria_day_2011(self):
        holidays = self.cal.holidays_set(2011)
        self.assertIn(date(2011, 5, 30), holidays)


class SpringHolidayFirstMondayJuneMixin(object):
    def test_spring_holiday(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 6, 2), holidays)


# -----------------------------------------------------------------------------
class ScotlandTest(GenericCalendarTest):
    cal_class = Scotland

    def test_year_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 1, 1), holidays)  # New year's day
        self.assertIn(date(2014, 1, 2), holidays)  # New year holiday
        self.assertIn(date(2014, 5, 5), holidays)  # May day
        self.assertIn(date(2014, 8, 4), holidays)  # Summer Holiday
        self.assertIn(date(2014, 11, 30), holidays)  # St. Andrew
        self.assertIn(date(2014, 12, 25), holidays)  # XMas
        self.assertIn(date(2014, 12, 26), holidays)  # Boxing day


class ScotlandInvernessTest(SpringHolidayFirstMondayAprilMixin, ScotlandTest):
    cal_class = ScotlandInverness

    def test_regional_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 2, 3), holidays)  # winter holiday
        self.assertIn(date(2014, 3, 3), holidays)  # 1st monday in march


class ScotlandLochaberTest(ScotlandTest):
    cal_class = ScotlandLochaber

    def test_regional_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 3, 31), holidays)  # last monday in march


class ScotlandAngusTest(SpringHolidaySecondMondayAprilMixin, ScotlandTest):
    cal_class = ScotlandAngus


class ScotlandAyrTest(
        GoodFridayMixin, EasterMondayMixin, SpringHolidayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandAyr


class ScotlandCarnoustieTest(SpringHolidayFirstMondayAprilMixin, ScotlandTest):
    cal_class = ScotlandCarnoustie


class ScotlandClydebankTest(SpringHolidayTuesdayMondayMayMixin, ScotlandTest):
    cal_class = ScotlandClydebank


class ScotlandDumfriesGallowayTest(GoodFridayMixin, ScotlandTest):
    cal_class = ScotlandDumfriesGalloway


class ScotlandEastDunbartonshireTest(
        GoodFridayMixin, EasterMondayMixin, SpringHolidayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandEastDunbartonshire


class ScotlandDundeeTest(
        SpringHolidayFirstMondayAprilMixin, VictoriaDayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandDundee


class ScotlandFifeTest(
        SpringHolidayFirstMondayAprilMixin, SpringHolidayFirstMondayJuneMixin,
        ScotlandTest):
    cal_class = ScotlandFife


class ScotlandEdinburghTest(
        GoodFridayMixin, EasterMondayMixin,
        VictoriaDayLastMondayBefore24MayMixin,
        ScotlandTest):
    cal_class = ScotlandEdinburgh

    def test_spring_holiday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 4, 14), holidays)  # Spring holiday, shifted

    def test_spring_holiday_2013(self):
        holidays = self.cal.holidays_set(2013)
        self.assertIn(date(2013, 4, 15), holidays)  # Spring holiday


class ScotlandElginTest(SpringHolidaySecondMondayAprilMixin, ScotlandTest):
    cal_class = ScotlandElgin


class ScotlandFalkirkTest(GoodFridayMixin, EasterMondayMixin, ScotlandTest):
    cal_class = ScotlandFalkirk


class ScotlandGalashiels(SpringHolidayFirstMondayJuneMixin, ScotlandTest):
    cal_class = ScotlandGalashiels


class ScotlandGlasgowTest(
        EasterMondayMixin, SpringHolidayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandGlasgow


class ScotlandInverclydeTest(
        GoodFridayMixin, EasterMondayMixin, SpringHolidayFirstMondayJuneMixin,
        ScotlandTest):
    cal_class = ScotlandInverclyde

    def test_spring_holiday_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 4, 28), holidays)  # Spring holiday


class ScotlandKilmarnockTest(GoodFridayMixin, EasterMondayMixin, ScotlandTest):
    cal_class = ScotlandKilmarnock


class ScotlandMonifiethTest(SpringHolidayFirstMondayAprilMixin, ScotlandTest):
    cal_class = ScotlandMonifieth


class ScotlandNorthLanarkshireTest(
        EasterMondayMixin, SpringHolidayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandNorthLanarkshire


class ScotlandPaisleyTest(
        GoodFridayMixin, EasterMondayMixin, VictoriaDayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandPaisley


class ScotlandPerthTest(SpringHolidayFirstMondayAprilMixin, ScotlandTest):
    cal_class = ScotlandPerth

    def test_victoria_day_2014(self):
        holidays = self.cal.holidays_set(2014)
        self.assertIn(date(2014, 5, 26), holidays)


class ScotlandScottishBordersTest(
        SpringHolidayFirstMondayAprilMixin,
        ScotlandTest):
    cal_class = ScotlandScottishBorders


class ScotlandStirlingTest(
        GoodFridayMixin, EasterMondayMixin, SpringHolidayTuesdayMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandStirling


class ScotlandSouthTest(GoodFridayMixin, ScotlandTest):
    cal_class = ScotlandSouth


class ScotlandSouthLanarkshireTest(
        GoodFridayMixin, EasterMondayMixin, SpringHolidayLastMondayMayMixin,
        ScotlandTest):
    cal_class = ScotlandSouthLanarkshire


class ScotlandWestDunbartonshireTest(
        GoodFridayMixin, EasterMondayMixin,
        ScotlandTest):
    cal_class = ScotlandWestDunbartonshire


workalendar/europe.py
from datetime import date, timedelta
from workalendar.core import WesternCalendar, ChristianMixin, OrthodoxMixin
from workalendar.core import THU, MON, FRI, SAT
from workalendar.scotland import *  # noqa


class CzechRepublic(WesternCalendar, ChristianMixin):
    "Czech Republic"
    include_easter_monday = True

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (1, 1, "Restoration Day of the Independent Czech State"),
        (5, 1, "Labour Day"),
        (5, 8, "Liberation Day"),
        (7, 5, "Saints Cyril and Methodius Day"),
        (7, 6, "Jan Hus Day"),
        (9, 28, "St. Wenceslas Day (Czech Statehood Day)"),
        (10, 28, "Independent Czechoslovak State Day"),
        (11, 17, "Struggle for Freedom and Democracy Day"),
        (12, 24, "Christmas Eve"),
        (12, 26, "St. Stephen's Day (The Second Christmas Day)"),
    )


class Sweden(WesternCalendar, ChristianMixin):
    "Sweden"
    include_epiphany = True
    include_good_friday = True
    include_easter_sunday = True
    include_easter_monday = True
    include_ascension = True
    include_whit_sunday = True
    whit_sunday_abel = "Pentecost"
    # Christmas Eve is not a holiday but not a work day either
    include_christmas_eve = True
    include_boxing_day = True
    boxing_day_label = "Second Day of Christmas"

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (5, 1, "Labour Day"),
        (6, 6, "National Day"),
        # New Year's Eve is not a holiday but not a work day either
        (12, 31, "New Year's Eve")
    )

    # Midsummer Eve is not a holiday but not a work day either
    def get_midsummer_eve(self, year):
        date_eve = Sweden.get_nth_weekday_in_month(
            year, 6, FRI, start=date(year, 6, 19))
        return date_eve

    def get_midsummer_day(self, year):
        date_eve = Sweden.get_nth_weekday_in_month(
            year, 6, SAT, start=date(year, 6, 20))
        return date_eve

    def get_variable_all_saints(self, year):
        all_saints = date(year, 10, 31)
        if all_saints.weekday() != SAT:
            all_saints = Sweden.get_nth_weekday_in_month(
                year, 11, SAT)
        return all_saints

    def get_variable_days(self, year):
        days = super(Sweden, self).get_variable_days(year)
        days.append((self.get_midsummer_day(year), "Midsummer's Day"))
        days.append((self.get_midsummer_eve(year), "Midsummer's Eve"))
        days.append((self.get_variable_all_saints(year), "All Saints"))
        return days


class Finland(WesternCalendar, ChristianMixin):
    "Finland"
    include_epiphany = True
    include_good_friday = True
    include_easter_sunday = True
    include_easter_monday = True
    include_ascension = True
    include_whit_sunday = True
    whit_sunday_label = 'Pentecost'
    include_christmas_eve = True
    include_boxing_day = True
    boxing_day_label = "St. Stephen's Day"

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (5, 1, "Labour Day"),
        (12, 6, "Labour Day"),
    )

    def get_midsummer_eve(self, year):
        date_eve = Finland.get_nth_weekday_in_month(
            year, 6, FRI, start=date(year, 6, 19))
        return date_eve

    def get_midsummer_day(self, year):
        date_eve = Finland.get_nth_weekday_in_month(
            year, 6, SAT, start=date(year, 6, 20))
        return date_eve

    def get_variable_all_saints(self, year):
        all_saints = date(year, 10, 31)
        if all_saints.weekday() != SAT:
            all_saints = Finland.get_nth_weekday_in_month(
                year, 11, SAT)
        return all_saints

    def get_variable_days(self, year):
        days = super(Finland, self).get_variable_days(year)
        days.append((self.get_midsummer_eve(year), "Midsummer's Eve"))
        days.append((self.get_midsummer_day(year), "Midsummer's Day"))
        days.append((self.get_variable_all_saints(year), "All Saints"))
        return days


class France(WesternCalendar, ChristianMixin):
    "France"
    include_easter_monday = True
    include_ascension = True
    include_whit_monday = True
    include_all_saints = True
    include_assumption = True

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (5, 1, "Labour Day"),
        (5, 8, "Victory in Europe Day"),
        (7, 14, "Bastille Day"),
        (11, 11, "Armistice Day"),
    )


class FranceAlsaceMoselle(France):
    "France Alsace/Moselle"
    include_good_friday = True
    include_boxing_day = True


class Greece(OrthodoxMixin, WesternCalendar):
    "Greece"
    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (3, 25, "Independence Day"),
        (5, 1, "Labour Day"),
        (10, 28, "Ohi Day"),
    )
    include_epiphany = True
    include_clean_monday = True
    include_annunciation = True
    include_good_friday = True
    include_easter_sunday = True
    include_easter_monday = True
    include_whit_sunday = True
    whit_sunday_label = "Pentecost"
    include_whit_monday = True
    include_assumption = True
    include_boxing_day = True
    boxing_day_label = "Glorifying Mother of God"


class Hungary(WesternCalendar, ChristianMixin):
    "Hungary"
    include_easter_sunday = True
    include_easter_monday = True
    include_whit_sunday = True
    whit_sunday_label = "Pentecost Sunday"
    include_whit_monday = True
    whit_monday_label = "Pentecost Monday"
    include_boxing_day = True
    boxing_day_label = "Second Day of Christmas"
    include_all_saints = True

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (3, 15, "National Day"),
        (5, 1, "Labour Day"),
        (8, 20, "St Stephen's Day"),
        (10, 23, "National Day"),
    )


class Iceland(WesternCalendar, ChristianMixin):
    "Iceland"
    include_holy_thursday = True
    include_good_friday = True
    include_easter_monday = True
    include_ascension = True
    include_whit_monday = True
    include_christmas_eve = True
    include_boxing_day = True
    boxing_day_label = "St Stephen's Day"

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (5, 1, "Labour Day"),
        (6, 17, "Icelandic National Day"),
        (12, 31, "New Year's Eve"),
    )

    def get_first_day_of_summer(self, year):
        """It's the first thursday *after* April, 18th.
        If April the 18th is a thursday, then it jumps to the 24th.
        """
        return WesternCalendar.get_nth_weekday_in_month(
            year, 4, THU,
            start=date(year, 4, 19))

    def get_variable_days(self, year):
        days = super(Iceland, self).get_variable_days(year)
        days += [
            (
                self.get_first_day_of_summer(year),
                "First day of summer"),
            (
                Iceland.get_nth_weekday_in_month(year, 8, MON),
                "Commerce Day"),
        ]
        return days


class Italy(WesternCalendar, ChristianMixin):
    "Italy"
    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (4, 25, "Liberation Day"),
        (5, 1, "International Workers' Day"),
        (6, 2, "Republic Day"),
    )
    include_immaculate_conception = True
    include_epiphany = True
    include_easter_monday = True
    include_assumption = True
    include_all_saints = True
    include_assumption = True
    include_boxing_day = True
    boxing_day_label = "St Stephen's Day"


class Norway(WesternCalendar, ChristianMixin):
    "Norway"
    include_holy_thursday = True
    include_good_friday = True
    include_easter_sunday = True
    include_easter_monday = True
    include_ascension = True
    include_whit_monday = True
    include_whit_sunday = True
    include_boxing_day = True
    boxing_day_label = "St Stephen's Day"

    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (5, 1, "Labour Day"),
        (5, 17, "Constitution Day"),
    )


class Poland(WesternCalendar, ChristianMixin):
    "Poland"
    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (1, 6, 'Trzech Kroli'),
        (5, 1, 'Labour Day'),
        (5, 3, 'Constitution Day'),
        (11, 11, 'Independence Day'),
    )
    include_easter_sunday = True
    include_easter_monday = True
    include_whit_sunday = True
    whit_sunday_label = "Pentecost Sunday"
    include_corpus_christi = True
    include_assumption = True
    include_all_saints = True
    include_boxing_day = True


class UnitedKingdom(WesternCalendar, ChristianMixin):
    "United Kingdom"
    include_good_friday = True
    include_easter_sunday = True
    include_easter_monday = True
    include_boxing_day = True
    shift_new_years_day = True

    def get_early_may_bank_holiday(self, year):
        return (
            UnitedKingdom.get_nth_weekday_in_month(year, 5, MON),
            "Early May Bank Holiday"
        )

    def get_spring_bank_holiday(self, year):
        return (
            UnitedKingdom.get_last_weekday_in_month(year, 5, MON),
            "Spring Bank Holiday"
        )

    def get_late_summer_bank_holiday(self, year):
        return (
            UnitedKingdom.get_last_weekday_in_month(year, 8, MON),
            "Late Summer Bank Holiday"
        )

    def get_variable_days(self, year):
        days = super(UnitedKingdom, self).get_variable_days(year)
        days.append(self.get_early_may_bank_holiday(year))
        days.append(self.get_spring_bank_holiday(year))
        days.append(self.get_late_summer_bank_holiday(year))
        # Boxing day & XMas shift
        christmas = date(year, 12, 25)
        if christmas.weekday() in self.get_weekend_days():
            shift = self.find_following_working_day(christmas)
            days.append((shift, "Christmas Shift"))
            days.append((shift + timedelta(days=1), "Boxing Day Shift"))
        return days


class UnitedKingdomNorthernIreland(UnitedKingdom):
    "Northern Ireland (UK)"
    def get_variable_days(self, year):
        days = super(UnitedKingdomNorthernIreland, self) \
            .get_variable_days(year)
        # St Patrick's day
        st_patrick = date(year, 3, 17)
        days.append((st_patrick, "Saint Patrick's Day"))
        if st_patrick.weekday() in self.get_weekend_days():
            days.append((
                self.find_following_working_day(st_patrick),
                "Saint Patrick substitute"))

        # Battle of boyne
        battle_of_boyne = date(year, 7, 12)
        days.append((battle_of_boyne, "Battle of the Boyne"))
        if battle_of_boyne.weekday() in self.get_weekend_days():
            days.append((
                self.find_following_working_day(battle_of_boyne),
                "Battle of the Boyne substitute"))
        return days


class EuropeanCentralBank(WesternCalendar, ChristianMixin):
    "European Central Bank"
    FIXED_HOLIDAYS = WesternCalendar.FIXED_HOLIDAYS + (
        (5, 1, "Labour Day"),
        (12, 26, "St. Stephen's Day"),
    )

    include_good_friday = True
    include_easter_monday = True
